### DISCUSSION TITLE
"Unblocking" py5.run_sketch() on MacOS

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#169] [Ideas]
I'm on Linux, and I'm enjoying running sketches with the `py5.run_sketch(block=False)` argument, that is unavailable on MacOS [for the standard Python interpreter, like in the Thonny IDE].

It allows me to inspect and fiddle with stuff in the REPL as the sketch is running. I'm pretty sure @ramalho would be delighted to use this, and I'd really like to be able to use this with students universally, that is,  on every OS.

![image](https://user-images.githubusercontent.com/3694604/193427766-ff278c37-fc04-46d0-a1c6-89dbdf648448.png)
https://twitter.com/py5coding/status/1567195642790805505?s=20&t=y3UBCXhXuVcDUWA8Re9ZZQ

The link mentioned in the tweet: https://github.com/py5coding/py5generator/blob/main/py5_resources/py5_module/py5/sketch.py#L194

I have opened this thread so that @osantana and @hx2A can track any progress on this, and maybe other people can join them without the formality of an "issue".

**Comment by hx2A** [comment]
> It allows me to inspect and fiddle with stuff in the REPL as the sketch is running. I'm pretty sure @ramalho would be delighted to use this, and I'd really like to be able to use this with students universally, that is, on every OS.

I'm a big fan of being able to continue to execute Python commands in the same environment as a running Sketch. This is wonderful if you are using Jupyter Notebook Widgets to control variables.

You do have the ability to inspect and fiddle with a running Sketch on OSX if you use Jupyter. Most commonly people use `jupyter notebook` to create the environment. There are other choices though, such as `jupyter console`. Most people don't know about that one, but it works in the terminal. There's also the `ipython` REPL, which is very similar to but not the same as `jupyter console`. Of course py5 has the blocking issue if you use the generic Python interpreter `python`. I recall that both `ipython` and `jupyter console` have some anomalies related to the use of `print()` or `py5.println()`? Nevertheless, there is a path forward here with Jupyter.

**Reply by villares** [reply]
I don't remember if we have another discussion thread for this but the example I have of a sketch with **multiple windows** depends on Class Mode and non-blocking py5.run_sketch(). 

So... would we have another multiple windows solution for MacOS sketches-not-on-jupyter-notebooks? Perhaps we could summon @ericof who is on MacOS to work on this...

**Reply by villares** [reply]
well I found it, here it is: https://github.com/py5coding/py5generator/discussions/158

**Reply by hx2A** [reply]
I remember that discussion!

I just tried this on my mac and I was able to run two Sketches simultaneously through Jupyter. This currently won't work with the Python terminal though. I had trouble exiting the second Sketch in Jupyter but they both ran correctly.

Multiple concurrent Sketches always requires class mode because you will have multiple instances of `py5.Sketch`.

Right now I want to get the next release out the door (the importing imported mode code feature) but once that is done I am willing to take another stab at this. Each time I work on it I make it better than it was before. Right now the `run_sketch()` code is intertwined with other stuff and might be harder for someone else to understand, but I have an idea for how to isolate the main issue which will make it easier for @ericof and others to understand and contribute to.
